src/lib/FreeC/Backend/Agda/Syntax.hs

```diff
@@ -70,8 +98,47 @@ intLiteral = Lit . LitNat NoRange
 lambda :: [Name] -> Expr -> Expr
 lambda args = Lam NoRange (DomainFree . defaultNamedArg . mkBinder_ <$> args)
 
--- | Creates an application AST node.
+-- | Creates an application AST node. Application is left associative and in
+--   in type expressions binds stronger than type arrow. For these cases paren-
+--   thesis are added automatically.
 --
---   @e a@
+--   > e a
 app :: Expr -> Expr -> Expr
-app l r = App NoRange l $ defaultNamedArg r
+app l r@(App _ _ _) = App NoRange l $ defaultNamedArg $ paren r
+app l r@(Fun _ _ _) = App NoRange l $ defaultNamedArg $ paren r
```

Comment:
Predicate with guard condition vs. code duplication. Not sure what I find more readable though. Any comments or preferences about that after calling it out?